src/propmachinery.jl

```diff
@@ -1,32 +1,95 @@
-import Base: start, next, done
-
 abstract QuPropagatorMethod
 
-immutable QuPropagator{QPM<:QuPropagatorMethod}
-    hamiltonian
-    init_state::QuVector
+abstract QuantumEquation
+
+immutable SchrodingerEquation{H<:QuBase.AbstractQuMatrix} <: QuantumEquation
+    hamiltonian::H
+    SchrodingerEquation(hamiltonian) = new(hamiltonian)
+end
+
+SchrodingerEquation{H<:QuBase.AbstractQuMatrix}(hamiltonian::H) = SchrodingerEquation{H}(hamiltonian)
+
+immutable LiouvillevonNeumannEquation{H<:QuBase.AbstractQuMatrix} <: QuantumEquation
+    liouvillian::H
+    LiouvillevonNeumannEquation(superoperator) = new(superoperator)
+end
+
+LiouvillevonNeumannEquation{H<:QuBase.AbstractQuMatrix}(liouvillian::H) = LiouvillevonNeumannEquation{H}(liouvillian)
+
+immutable QuPropagator{QPM<:QuPropagatorMethod, QVM<:Union(QuBase.AbstractQuVector,QuBase.AbstractQuMatrix)}
+    eq
+    init_state::QVM
     tlist
     method::QPM
-    QuPropagator(hamiltonian, init_state, tlist, method) = new(hamiltonian, init_state, tlist, method)
+    QuPropagator(eq, init_state, tlist, method) = new(eq, init_state, tlist, method)
 end
 
-QuPropagator{QPM<:QuPropagatorMethod}(hamiltonian, init_state, tlist, method::QPM) = QuPropagator{QPM}(hamiltonian, init_state, tlist, method)
+QuPropagator{QPM<:QuPropagatorMethod, QV<:QuBase.AbstractQuVector}(eq::SchrodingerEquation, init_state::QV, tlist, method::QPM) = QuPropagator{QPM,QV}(eq, init_state, tlist, method)
+
+QuPropagator{QPM<:QuPropagatorMethod, QV<:QuBase.AbstractQuVector}(param::QuBase.AbstractQuMatrix, init_state::QV,  tlist, method::QPM) = QuPropagator{QPM,QV}(SchrodingerEquation(param),init_state, tlist, method)
+
+QuPropagator{QPM<:QuPropagatorMethod, QM<:QuBase.AbstractQuMatrix}(eq::LiouvillevonNeumannEquation, init_state::QM, tlist, method::QPM) = QuPropagator{QPM,QM}(eq, init_state, tlist, method)
+
+QuPropagator{QPM<:QuPropagatorMethod, QM<:QuBase.AbstractQuMatrix}(param::QuBase.AbstractQuMatrix, init_state::QM,  tlist, method::QPM) = QuPropagator{QPM,QM}(LiouvillevonNeumannEquation(liouvillian_op(param)),init_state, tlist, method)
 
 immutable QuPropagatorState
-    psi::QuVector
+    psi
     t
     t_state
 end
 
+function operator(qu_eq::LiouvillevonNeumannEquation)
+    return qu_eq.liouvillian
+end
+
+function operator(qu_eq::SchrodingerEquation)
+    return qu_eq.hamiltonian
+end
+
+function liouvillian_op(h::QuBase.AbstractQuMatrix)
+    nb = size(coeffs(h), 1)
+    SI = Array(Int,0)
+    SJ = Array(Int,0)
+    Lvals = Array(Complex128,0)
```

Comment:
Here we will need something like `typeof(im*one(eltype(h)))` instead of `Complex128` to be general.